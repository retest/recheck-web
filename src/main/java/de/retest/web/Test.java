package de.retest.web;

import java.util.List;

public class Test {

	public boolean getFoo( final List<String> list ) {

		boolean result = false;

		for ( final String s : list ) {
			if ( s.equals( "Test" ) ) {
				result = true;
			}
		}

		return result;
	}

}
