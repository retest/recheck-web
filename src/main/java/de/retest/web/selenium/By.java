package de.retest.web.selenium;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

import de.retest.recheck.ui.descriptors.Element;
import de.retest.recheck.ui.descriptors.RootElement;
import de.retest.recheck.ui.diff.Alignment;

public abstract class By extends org.openqa.selenium.By {

	public static ByBestMatchToRetestId retestId( final String retestId ) {
		return new ByBestMatchToRetestId( retestId );
	}

	public static Element findElement( final RootElement lastExpectedState, final RootElement lastActualState,
			final Predicate<Element> predicate ) {
		if ( lastExpectedState == null ) {
			throw new IllegalArgumentException( "Cannot find element in null state." );
		}
		final Element resultFromExpected = findElement( lastExpectedState.getContainedElements(), predicate );
		if ( resultFromExpected == null ) {
			return null;
		}
		final Alignment alignment = Alignment.createAlignment( lastExpectedState, lastActualState );
		final Element resultFromActual = alignment.getActual( resultFromExpected );
		if ( resultFromActual == null ) {
			throw new NoElementWithHighEnoughMatchFoundException( resultFromExpected );
		}
		return resultFromActual.applyRetestId( resultFromExpected.getRetestId() );
	}

	public static Element findElement( final List<Element> children, final Predicate<Element> predicate ) {
		for ( final Element element : children ) {
			if ( predicate.test( element ) ) {
				return element;
			}
			final Element result = findElement( element.getContainedElements(), predicate );
			if ( result != null ) {
				return result;
			}
		}
		return null;
	}

	public static List<Element> findElements( final List<Element> children, final Predicate<Element> predicate ) {
		final List<Element> result = new ArrayList<>();
		for ( final Element element : children ) {
			if ( predicate.test( element ) ) {
				result.add( element );
			}
			result.addAll( findElements( element.getContainedElements(), predicate ) );
		}
		return result;
	}

	public static Element findElementByAttribute( final RootElement lastExpectedState,
			final RootElement lastActualState, final String attributeName, final Predicate<Object> condition ) {
		return findElement( lastExpectedState, lastActualState, element -> {
			if ( element.getIdentifyingAttributes().get( attributeName ) != null ) {
				return condition.test( element.getIdentifyingAttributes().getAttribute( attributeName ).getValue() );
			} else if ( element.getAttributes().get( attributeName ) != null ) {
				return condition.test( element.getAttributes().get( attributeName ) );
			} else {
				return false;
			}
		} );
	}

	public static Element findElementByAttribute( final RootElement lastExpectedState,
			final RootElement lastActualState, final String attributeName, final Object attributeValue ) {
		return findElementByAttribute( lastExpectedState, lastActualState, attributeName, attributeValue::equals );
	}

}
