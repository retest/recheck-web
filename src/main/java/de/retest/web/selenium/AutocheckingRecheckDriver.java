package de.retest.web.selenium;

import java.util.List;
import java.util.stream.Collectors;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;

import de.retest.web.RecheckWebImpl;

public class AutocheckingRecheckDriver extends RecheckDriver {

	private RecheckWebImpl re;

	public AutocheckingRecheckDriver( final RemoteWebDriver wrapped ) {
		super( wrapped );
	}

	public void startTest() {
		counter = 1;
		if ( re == null ) {
			re = new RecheckWebImpl();
		}
		re.startTest();
	}

	public void startTest( final String testName ) {
		if ( re == null ) {
			re = new RecheckWebImpl();
		}
		re.startTest( testName );
	}

	public void capTest() {
		check( "final" );
		counter = 1;
		re.capTest();
	}

	public void cap() {
		counter = 1;
		re.cap();
	}

	@Override
	public void close() {
		// Is this sensible? What about tests using separate sessions?
		counter = 1;
		re.cap();
		super.close();
	}

	@Override
	public void quit() {
		// Is this sensible? What about tests using separate sessions?
		counter = 1;
		re.cap();
		super.quit();
	}

	@Override
	public WebElement findElement( final ByBestMatchToRetestId by ) {
		return new WebElementWrapper( super.findElement( by ), this );
	}

	@Override
	public WebElement findElement( final By by ) {
		return new WebElementWrapper( super.findElement( by ), this );
	}

	@Override
	public List<WebElement> findElements( final By by ) {
		return super.findElements( by ).stream().map( element -> new WebElementWrapper( element, this ) )
				.collect( Collectors.toList() );
	}

	public void check( final String currentStep ) {
		re.check( this, makeUnique( currentStep ) );
	}

	private int counter = 1;

	String makeUnique( final String id ) {
		return String.format( "%02d_%s", counter++, id );
	}
}
