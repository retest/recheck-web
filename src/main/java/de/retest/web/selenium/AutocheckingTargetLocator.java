package de.retest.web.selenium;

import org.openqa.selenium.Alert;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import lombok.RequiredArgsConstructor;
import org.openqa.selenium.WindowType;

@RequiredArgsConstructor
public class AutocheckingTargetLocator implements WebDriver.TargetLocator {

	private final WebDriver.TargetLocator delegate;
	private final AutocheckingRecheckDriver driver;

	@Override
	public WebDriver frame( final int index ) {
		return delegate.frame( index );
	}

	@Override
	public WebDriver frame( final String nameOrId ) {
		return delegate.frame( nameOrId );
	}

	@Override
	public WebDriver frame( final WebElement frameElement ) {
		return delegate.frame( frameElement );
	}

	@Override
	public WebDriver parentFrame() {
		return delegate.parentFrame();
	}

	@Override
	public WebDriver window( final String nameOrHandle ) {
		final WebDriver driver = delegate.window( nameOrHandle );
		this.driver.check( "switch-window" );
		return driver;
	}

	@Override
	public WebDriver newWindow(WindowType typeHint) {
		return delegate.newWindow(typeHint);
	}

	@Override
	public WebDriver defaultContent() {
		return delegate.defaultContent();
	}

	@Override
	public WebElement activeElement() {
		return driver.wrap( delegate.activeElement() ) ;
	}

	@Override
	public Alert alert() {
		return delegate.alert();
	}
}
