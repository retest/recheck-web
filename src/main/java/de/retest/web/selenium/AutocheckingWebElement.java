package de.retest.web.selenium;

import java.util.List;
import java.util.stream.Collectors;

import org.openqa.selenium.By;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.Point;
import org.openqa.selenium.Rectangle;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.WrapsDriver;
import org.openqa.selenium.interactions.Coordinates;
import org.openqa.selenium.interactions.Locatable;
import org.openqa.selenium.internal.WrapsElement;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor( access = AccessLevel.PRIVATE )
public class AutocheckingWebElement implements WebElement, WrapsDriver, WrapsElement, Locatable {

	private final WebElement wrappedElement;
	private final AutocheckingRecheckDriver driver;

	public static AutocheckingWebElement of( final WebElement wrappedElement, final AutocheckingRecheckDriver driver ) {
		if ( wrappedElement instanceof AutocheckingWebElement ) {
			return (AutocheckingWebElement) wrappedElement;
		}
		return new AutocheckingWebElement( wrappedElement, driver );
	}

	public static AutocheckingWebElement of( final WebElement wrappedElement, final AutocheckingRecheckDriver driver,
			final String representation ) {
		if ( wrappedElement instanceof AutocheckingWebElement ) {
			return (AutocheckingWebElement) wrappedElement;
		}

		return new AutocheckingWebElement( wrappedElement, driver ) {
			@Override
			public String toString() {
				return representation;
			}
		};
	}

	/**
	 * Skip checks for actions performed on this web element.
	 *
	 * @return returns the plain WebElement
	 */
	public WebElement skipCheck() {
		return wrappedElement;
	}

	@Override
	public <X> X getScreenshotAs( final OutputType<X> target ) throws WebDriverException {
		return wrappedElement.getScreenshotAs( target );
	}

	@Override
	public void click() {
		wrappedElement.click();
		driver.check( "click", wrappedElement );
	}

	@Override
	public void submit() {
		wrappedElement.submit();
		driver.check( "submit", wrappedElement );
	}

	@Override
	public void sendKeys( final CharSequence... keysToSend ) {
		wrappedElement.sendKeys( keysToSend );
		driver.check( "enter", wrappedElement, (Object[]) keysToSend );
	}

	@Override
	public void clear() {
		wrappedElement.clear();
		driver.check( "clear", wrappedElement );
	}

	@Override
	public String getTagName() {
		return wrappedElement.getTagName();
	}

	@Override
	public String getAttribute( final String name ) {
		return wrappedElement.getAttribute( name );
	}

	@Override
	public boolean isSelected() {
		return wrappedElement.isSelected();
	}

	@Override
	public boolean isEnabled() {
		return wrappedElement.isEnabled();
	}

	@Override
	public String getText() {
		return wrappedElement.getText();
	}

	@Override
	public List<WebElement> findElements( final By by ) {
		return wrappedElement.findElements( by ).stream() //
				.map( element -> of( element, driver ) ) //
				.collect( Collectors.toList() );
	}

	@Override
	public WebElement findElement( final By by ) {
		return of( wrappedElement.findElement( by ), driver );
	}

	@Override
	public boolean isDisplayed() {
		return wrappedElement.isDisplayed();
	}

	@Override
	public Point getLocation() {
		return wrappedElement.getLocation();
	}

	@Override
	public Dimension getSize() {
		return wrappedElement.getSize();
	}

	@Override
	public Rectangle getRect() {
		return wrappedElement.getRect();
	}

	@Override
	public String getCssValue( final String propertyName ) {
		return wrappedElement.getCssValue( propertyName );
	}

	@Override
	public WebDriver getWrappedDriver() {
		return driver;
	}

	@Override
	public WebElement getWrappedElement() {
		return wrappedElement;
	}

	@Override
	public String toString() {
		return wrappedElement.toString();
	}

	@Override
	public Coordinates getCoordinates() {
		if ( wrappedElement instanceof Locatable ) {
			return ((Locatable) wrappedElement).getCoordinates();
		}
		throw new IllegalStateException(
				String.format( "Element is not instance of %s.", Locatable.class.getSimpleName() ) );
	}
}
