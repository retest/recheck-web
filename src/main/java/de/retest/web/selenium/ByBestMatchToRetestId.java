package de.retest.web.selenium;

import java.io.Serializable;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.retest.recheck.ui.descriptors.Element;
import de.retest.recheck.ui.descriptors.RootElement;

/**
 * {@code ByBestMatchToRetestId} will search for elements by their old retest ID (the retest ID they had in the Golden
 * Master state).
 *
 * Note that elements are assigned a retest ID when they are created (i.e. also when the <em>new</em> state is created).
 * This retest ID might differ from the retest ID the element had in the persisted (Golden Master) state. Therefore, all
 * its child elements have their respective <em>new</em> (potentially different) retest ID. Thus, the result of a call
 * to {@code ByBestMatchToRetestId} might return an element with children that have a different retest ID than they have
 * in the Golden Master.
 */
public class ByBestMatchToRetestId extends By implements Serializable {

	private static final long serialVersionUID = -3787115401615364934L;

	private static final Logger logger = LoggerFactory.getLogger( ByBestMatchToRetestId.class );

	private final String retestId;

	public ByBestMatchToRetestId( final String retestId ) {
		this.retestId = retestId;
		if ( StringUtils.isBlank( retestId ) ) {
			throw new IllegalArgumentException( "retestId must not be empty." );
		}
	}

	@Override
	public List<WebElement> findElements( final SearchContext context ) {
		logger.info( "findElements called with {} of {}.", context, context.getClass() );
		throw new UnsupportedOperationException( "This class can only be used in conjunction with a RecheckDriver." );
	}

	/**
	 * @param lastExpectedState
	 *            The last expected state (usually the Golden Master).
	 * @param lastActualState
	 *            The last actual state (usually the current state).
	 * @return Maybe an element whose children have a different retest ID than in the Golden Master.
	 */
	public Element findElement( final RootElement lastExpectedState, final RootElement lastActualState ) {
		if ( lastExpectedState == null ) {
			// find by retestId even in case of just creating the state
			return de.retest.web.selenium.By.findElement( lastActualState.getContainedElements(),
					element -> retestId.equals( element.getRetestId() ) );
		}
		final Element result = de.retest.web.selenium.By.findElement( lastExpectedState, lastActualState,
				element -> retestId.equals( element.getRetestId() ) );
		if ( result == null ) {
			throw new NoElementWithReTestIdFoundException( retestId );
		}
		return result;
	}

	@Override
	public int hashCode() {
		return retestId.hashCode();
	}

	@Override
	public boolean equals( final Object other ) {
		if ( other == null ) {
			return false;
		}
		if ( this == other ) {
			return true;
		}
		if ( getClass() != other.getClass() ) {
			return false;
		}
		return retestId.equals( ((ByBestMatchToRetestId) other).retestId );
	}

	public String getRetestId() {
		return retestId;
	}
}
