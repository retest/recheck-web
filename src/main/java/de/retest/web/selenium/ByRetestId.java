package de.retest.web.selenium;

import java.io.Serializable;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.retest.ui.descriptors.Element;
import de.retest.ui.descriptors.RootElement;
import de.retest.ui.diff.Alignment;

public class ByRetestId extends By implements Serializable {

	private static final long serialVersionUID = -3787115401615364934L;

	private static final Logger logger = LoggerFactory.getLogger( ByRetestId.class );

	private final String retestId;

	public ByRetestId( final String retestId ) {
		this.retestId = retestId;
		if ( StringUtils.isBlank( retestId ) ) {
			throw new IllegalArgumentException( "retestId must not be empty." );
		}
	}

	@Override
	public List<WebElement> findElements( final SearchContext context ) {
		logger.info( "findElements called with {} of {}.", context, context.getClass() );
		throw new UnsupportedOperationException( "This class can only be used in conjunction with a RecheckDriver." );
	}

	public Element findElement( final RootElement lastExpectedState, final RootElement lastActualState ) {
		if ( lastExpectedState == null ) {
			throw new IllegalArgumentException( "Cannot find element in null state." );
		}
		final Element result = findElement( lastExpectedState.getContainedElements() );
		if ( result != null ) {
			// TODO Use unobfuscated methods
			final Alignment alignment = Alignment.a( lastExpectedState, lastActualState );
			return alignment.a( result );
		}
		throw new RuntimeException( "No element with retest id '" + retestId + "' found!" );
	}

	public Element findElement( final List<Element> children ) {
		for ( final Element element : children ) {
			if ( retestId.equals( element.getRetestId() ) ) {
				return element;
			}
			final Element result = findElement( element.getContainedElements() );
			if ( result != null ) {
				return result;
			}
		}
		return null;
	}
}
