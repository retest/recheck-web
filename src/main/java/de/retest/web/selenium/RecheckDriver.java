package de.retest.web.selenium;

import java.util.List;
import java.util.Set;

import org.openqa.selenium.By;
import org.openqa.selenium.Capabilities;
import org.openqa.selenium.HasCapabilities;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.html5.LocalStorage;
import org.openqa.selenium.html5.Location;
import org.openqa.selenium.html5.LocationContext;
import org.openqa.selenium.html5.SessionStorage;
import org.openqa.selenium.html5.WebStorage;
import org.openqa.selenium.interactions.HasInputDevices;
import org.openqa.selenium.interactions.HasTouchScreen;
import org.openqa.selenium.interactions.Keyboard;
import org.openqa.selenium.interactions.Mouse;
import org.openqa.selenium.interactions.TouchScreen;
import org.openqa.selenium.internal.FindsByClassName;
import org.openqa.selenium.internal.FindsByCssSelector;
import org.openqa.selenium.internal.FindsById;
import org.openqa.selenium.internal.FindsByLinkText;
import org.openqa.selenium.internal.FindsByName;
import org.openqa.selenium.internal.FindsByTagName;
import org.openqa.selenium.internal.FindsByXPath;

import de.retest.ui.descriptors.Element;
import de.retest.ui.descriptors.RootElement;

public class RecheckDriver implements WebDriver, JavascriptExecutor, FindsById, FindsByClassName, FindsByLinkText,
		FindsByName, FindsByCssSelector, FindsByTagName, FindsByXPath, HasInputDevices, HasCapabilities,
		TakesScreenshot, LocationContext, WebStorage, HasTouchScreen {

	private final WebDriver wrapped;
	private RootElement lastExpectedState;
	private RootElement lastActualState;

	public RecheckDriver( final WebDriver wrapped ) {
		this.wrapped = wrapped;
	}

	public void setLastExpectdState( final RootElement lastExpectedState ) {
		this.lastExpectedState = lastExpectedState;
	}

	public void setLastActualState( final RootElement lastActualState ) {
		this.lastActualState = lastActualState;
	}

	public WebElement findElement( final ByRetestId by ) {
		if ( lastExpectedState == null ) {
			throw new IllegalStateException(
					"You must first check the state before being able to withdraw elements from it." );
		}
		final Element searchedFor = by.findElement( lastExpectedState, lastActualState );
		return wrapped.findElement( By.xpath( searchedFor.getIdentifyingAttributes().getPath() ) );
	}

	public WebElement findElementByRecheckId( final String recheckId ) {
		return findElement( new ByRetestId( recheckId ) );
	}

	@Override
	public WebElement findElement( final By by ) {
		if ( by instanceof ByRetestId ) {
			return findElement( (ByRetestId) by );
		}
		// TODO We can implement to find the element with the given value in the lastCheckedState
		// Should it differ, we can use the new value in the current state and issue a warning!
		return wrapped.findElement( by );
	}

	@Override
	public List<WebElement> findElements( final By by ) {
		return wrapped.findElements( by );
	}

	@Override
	public WebElement findElementById( final String using ) {
		return ((FindsById) wrapped).findElementById( using );
	}

	@Override
	public List<WebElement> findElementsById( final String using ) {
		return ((FindsById) wrapped).findElementsById( using );
	}

	@Override
	public WebElement findElementByClassName( final String using ) {
		return ((FindsByClassName) wrapped).findElementByClassName( using );
	}

	@Override
	public List<WebElement> findElementsByClassName( final String using ) {
		return ((FindsByClassName) wrapped).findElementsByClassName( using );
	}

	@Override
	public WebElement findElementByLinkText( final String using ) {
		return ((FindsByLinkText) wrapped).findElementByLinkText( using );
	}

	@Override
	public List<WebElement> findElementsByLinkText( final String using ) {
		return ((FindsByLinkText) wrapped).findElementsByLinkText( using );
	}

	@Override
	public WebElement findElementByPartialLinkText( final String using ) {
		return ((FindsByLinkText) wrapped).findElementByPartialLinkText( using );
	}

	@Override
	public List<WebElement> findElementsByPartialLinkText( final String using ) {
		return ((FindsByLinkText) wrapped).findElementsByPartialLinkText( using );
	}

	@Override
	public WebElement findElementByName( final String using ) {
		return ((FindsByName) wrapped).findElementByName( using );
	}

	@Override
	public List<WebElement> findElementsByName( final String using ) {
		return ((FindsByName) wrapped).findElementsByName( using );
	}

	@Override
	public WebElement findElementByCssSelector( final String using ) {
		return ((FindsByCssSelector) wrapped).findElementByCssSelector( using );
	}

	@Override
	public List<WebElement> findElementsByCssSelector( final String using ) {
		return ((FindsByCssSelector) wrapped).findElementsByCssSelector( using );
	}

	@Override
	public WebElement findElementByTagName( final String using ) {
		return ((FindsByTagName) wrapped).findElementByTagName( using );
	}

	@Override
	public List<WebElement> findElementsByTagName( final String using ) {
		return ((FindsByTagName) wrapped).findElementsByTagName( using );
	}

	@Override
	public WebElement findElementByXPath( final String using ) {
		return ((FindsByXPath) wrapped).findElementByXPath( using );
	}

	@Override
	public List<WebElement> findElementsByXPath( final String using ) {
		return ((FindsByXPath) wrapped).findElementsByXPath( using );
	}

	@Override
	public void close() {
		wrapped.close();
	}

	@Override
	public void get( final String url ) {
		wrapped.get( url );
	}

	@Override
	public String getCurrentUrl() {
		return wrapped.getCurrentUrl();
	}

	@Override
	public String getPageSource() {
		return wrapped.getPageSource();
	}

	@Override
	public String getTitle() {
		return wrapped.getTitle();
	}

	@Override
	public String getWindowHandle() {
		return wrapped.getWindowHandle();
	}

	@Override
	public Set<String> getWindowHandles() {
		return wrapped.getWindowHandles();
	}

	@Override
	public Options manage() {
		return wrapped.manage();
	}

	@Override
	public Navigation navigate() {
		return wrapped.navigate();
	}

	@Override
	public void quit() {
		wrapped.quit();
	}

	@Override
	public TargetLocator switchTo() {
		return wrapped.switchTo();
	}

	@Override
	public Object executeScript( final String script, final Object... args ) {
		return ((JavascriptExecutor) wrapped).executeScript( script, args );
	}

	@Override
	public Object executeAsyncScript( final String script, final Object... args ) {
		return ((JavascriptExecutor) wrapped).executeAsyncScript( script, args );
	}

	@Override
	public <X> X getScreenshotAs( final OutputType<X> target ) throws WebDriverException {
		return ((TakesScreenshot) wrapped).getScreenshotAs( target );
	}

	@Override
	public Keyboard getKeyboard() {
		return ((HasInputDevices) wrapped).getKeyboard();
	}

	@Override
	public Mouse getMouse() {
		return ((HasInputDevices) wrapped).getMouse();
	}

	@Override
	public Capabilities getCapabilities() {
		return ((HasCapabilities) wrapped).getCapabilities();
	}

	@Override
	public Location location() {
		return ((LocationContext) wrapped).location();
	}

	@Override
	public void setLocation( final Location location ) {
		((LocationContext) wrapped).setLocation( location );
	}

	@Override
	public LocalStorage getLocalStorage() {
		return ((WebStorage) wrapped).getLocalStorage();
	}

	@Override
	public SessionStorage getSessionStorage() {
		return ((WebStorage) wrapped).getSessionStorage();
	}

	@Override
	public TouchScreen getTouch() {
		return ((HasTouchScreen) wrapped).getTouch();
	}
}
