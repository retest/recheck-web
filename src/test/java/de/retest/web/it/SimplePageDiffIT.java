package de.retest.web.it;

import static org.junit.jupiter.api.Assertions.fail;

import org.approvaltests.Approvals;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.WebDriver;

import de.retest.recheck.Recheck;
import de.retest.recheck.RecheckImpl;
import de.retest.web.testutils.PageFactory;
import de.retest.web.testutils.WebDriverFactory;
import de.retest.web.testutils.WebDriverFactory.Driver;

public class SimplePageDiffIT {

	WebDriver driver;
	Recheck re;

	@Before
	public void setup() {
		re = new RecheckImpl();
		driver = WebDriverFactory.driver( Driver.CHROME );
	}

	@Test
	public void testSimpleChange() throws Exception {
		re.startTest();

		driver.get( PageFactory.toPageUrlString( "simple-page-diff.html" ) );

		re.check( driver, "open" );
		try {
			re.capTest();
			fail( "Assertion Error expected" );
		} catch ( final AssertionError e ) {
			Approvals.verify( e.getMessage() );
		}
	}

	@After
	public void tearDown() {
		driver.quit();
		re.cap();
	}
}
