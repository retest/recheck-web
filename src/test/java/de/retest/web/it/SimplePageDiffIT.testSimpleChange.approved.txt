A detailed report will be created at 'C:\Users\Pascal Allion\eclipse-workspace\recheck-web\target\test-classes\retest\recheck\de.retest.web.it.SimplePageDiffIT.report'. You can review the details by using our CLI (https://github.com/retest/recheck.cli/) or GUI (https://retest.de/review/).

1 check(s) in 'de.retest.web.it.SimplePageDiffIT' found the following difference(s):
Test 'testSimpleChange' has 4 difference(s) in 1 state(s):
open resulted in:
	div at 'html[1]/body[1]/div[3]':
		id: expected="twoblocks", actual="changedblock"
	p [Some text] at 'html[1]/body[1]/div[3]/p[1]':
		text: expected="Some text", actual="Some changed text"	html[1]/body[1]/div[3]/p[2] was deleted!
	html[1]/body[1]/h2[1] was inserted!
